test/controller.go

```diff
@@ -0,0 +1,220 @@
+	"knative.dev/pkg/controller"
+)
+
+// Data represents the desired state of the system (i.e. existing resources) to seed controllers
+// with.
+type Data struct {
+	ResolutionRequests []*v1alpha1.ResolutionRequest
+	Namespaces         []*corev1.Namespace
+	ConfigMaps         []*corev1.ConfigMap
+	ServiceAccounts    []*corev1.ServiceAccount
+	LimitRange         []*corev1.LimitRange
+}
+
+// Clients holds references to clients which are useful for reconciler tests.
+type Clients struct {
+	ResolutionRequests *fakeresolutionclientset.Clientset
+	Kube               *fakekubeclientset.Clientset
+}
+
+// Informers holds references to informers which are useful for reconciler tests.
+type Informers struct {
+	ConfigMap         coreinformers.ConfigMapInformer
+	ServiceAccount    coreinformers.ServiceAccountInformer
+	LimitRange        coreinformers.LimitRangeInformer
+	ResolutionRequest resolutioninformersv1alpha1.ResolutionRequestInformer
+}
+
+// Assets holds references to the controller, logs, clients, and informers.
+type Assets struct {
+	Logger     *zap.SugaredLogger
+	Controller *controller.Impl
+	Clients    Clients
+	Informers  Informers
+	Recorder   *record.FakeRecorder
+	Ctx        context.Context
+}
+
+// AddToInformer returns a function to add ktesting.Actions to the cache store
+func AddToInformer(t *testing.T, store cache.Store) func(ktesting.Action) (bool, runtime.Object, error) {
+	return func(action ktesting.Action) (bool, runtime.Object, error) {
+		switch a := action.(type) {
+		case ktesting.CreateActionImpl:
+			if err := store.Add(a.GetObject()); err != nil {
+				t.Fatal(err)
+			}
+
+		case ktesting.UpdateActionImpl:
+			objMeta, err := meta.Accessor(a.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+
+			// Look up the old copy of this resource and perform the optimistic concurrency check.
+			old, exists, err := store.GetByKey(objMeta.GetNamespace() + "/" + objMeta.GetName())
+			if err != nil {
+				return true, nil, err
+			} else if !exists {
+				// Let the client return the error.
+				return false, nil, nil
+			}
+			oldMeta, err := meta.Accessor(old)
+			if err != nil {
+				return true, nil, err
+			}
+			// If the resource version is mismatched, then fail with a conflict.
+			if oldMeta.GetResourceVersion() != objMeta.GetResourceVersion() {
+				return true, nil, apierrs.NewConflict(
+					a.Resource.GroupResource(), objMeta.GetName(),
+					fmt.Errorf("resourceVersion mismatch, got: %v, wanted: %v",
+						objMeta.GetResourceVersion(), oldMeta.GetResourceVersion()))
+			}
+
+			// Update the store with the new object when it's fine.
+			if err := store.Update(a.GetObject()); err != nil {
+				t.Fatal(err)
+			}
+		}
+		return false, nil, nil
+	}
+}
+
+// SeedTestData returns Clients and Informers populated with the
+// given Data.
+// nolint: revive
+func SeedTestData(t *testing.T, ctx context.Context, d Data) (Clients, Informers) {
```

Comment:
This is basically a stripped down copy of Pipeline's `test/controller.go`, and will go away once we combine the repos.